### PR TITLE
KAFKA-2562: update kafka scripts to use new tools/code

### PR DESCRIPTION
**Pull request by omkreddy**
Updated  kafka-producer-perf-test.sh to use org.apache.kafka.clients.tools.ProducerPerformance. 
Updated build.gradle to add kafka-tools-0.9.0.0-SNAPSHOT.jar to kafka/libs  folder.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -244,6 +244,7 @@ project(':core') {
   dependencies {
     compile project(':clients')
     compile project(':log4j-appender')
+    compile project(':tools')
```
Comment (ijuma) [inline build.gradle]: Why is this needed?

Comment (ijuma) [inline build.gradle]: Oh, it's to add it to `kafka//libs` folder from your description. It seems a bit strange for `core` to depend on `tools` though.

File: bin/kafka-producer-perf-test.sh [hunk 2]
```diff
@@ -17,4 +17,4 @@
 if [ "x$KAFKA_HEAP_OPTS" = "x" ]; then
     export KAFKA_HEAP_OPTS="-Xmx512M"
 fi
-exec $(dirname $0)/kafka-run-class.sh kafka.tools.ProducerPerformance $@
+exec $(dirname $0)/kafka-run-class.sh org.apache.kafka.clients.tools.ProducerPerformance $@
```
Comment (ijuma) [inline bin/kafka-producer-perf-test.sh]: The new `ProducerPerformance` command is not compatible with the old one. Do we offer compatibility guarantees for these scripts? cc @junrao @ewencp

File: build.gradle (hunk 1 above)
Comment (omkreddy) [inline build.gradle]: Yes, Its bit odd.  Currently we are building kafka tar build from core sub-project. So we need to add this dependency to bring kafka-tools.jar to libs directory.  We can change this gradle build logic. I will try to fix this issue in next patch.

Comment (omkreddy) [inline build.gradle]: removed unnecessary project dependencies.

Comment (ijuma) [inline build.gradle]: Thanks.

File: build.gradle [hunk 3]
```diff
@@ -240,7 +240,7 @@ project(':core') {
 
   dependencies {
     compile project(':clients')
-    compile project(':log4j-appender')
+    compile "$slf4jlog4j"
```
Comment (omkreddy) [inline build.gradle:255]: Added slf4j, log4j jar dependency to core project.  This is to avoid old version of transitive dependencies from Zookeeper, ZKClient dependencies.  ZK and ZKClient depends on older version of slf4j and log4j.

File: bin/kafka-producer-perf-test.sh (hunk 2 above)
Comment (junrao) [inline bin/kafka-producer-perf-test.sh]: The old ProducerPerformance mixes performance testing and system testing together. The logic is now split into  VerifiableProducer and the new ProducerPerformance. So the command line won't be compatible.

File: build.gradle (hunk 3 above)
Comment (junrao) [inline build.gradle:255]: This is unfortunate. Ideally, Zookeeper shouldn't specify explicitly dependency on slf4j-log4j. We can keep the changes you have here.

**Comment by junrao** [comment]
Also, since we have moved to java 7, we probably should switch to the G1 garbage collector in kafka-run-class.sh. We can probably use sth like the following.

-XX:PermSize=96m -XX:MaxPermSize=96m -XX:+UseG1GC -XX:MaxGCPauseMillis=20 -XX:InitiatingHeapOccupancyPercent=35

File: bin/kafka-producer-perf-test.sh (hunk 2 above)
Comment (ijuma) [inline bin/kafka-producer-perf-test.sh]: Right, but should we be enabling the new one only if a config is passed? Or alternatively, use the old one if a config is passed. Maybe not worth it, I don't know if people rely on these scripts for something important.

File: build.gradle (hunk 3 above)
Comment (ijuma) [inline build.gradle:255]: Yeah, would you mind filing an issue with them? cc @fpj

**Comment by ijuma** [comment]
@junrao The change you are asking for is in another PR https://github.com/apache/kafka/pull/218

File: bin/kafka-producer-perf-test.sh (hunk 2 above)
Comment (junrao) [inline bin/kafka-producer-perf-test.sh]: The new ProducerPerformance is only going to support the new producer while the old producer supports both the old and the new producer. Also, we are not removing the old ProducerPerformance class. So, one can still use it if needed. Given that, I don't feel that we need to preserve the command line options. However, the command line option for the new ProducerPerformance is weird since it doesn't use jopt. The reason is probably because that class was in clients and we didn't want to drag in the jopt dependency to clients. Now that the new ProducerPerformance is in tools, we can change the command line options to use jopt so that it's consistent with other tools.

Comment (ijuma) [inline bin/kafka-producer-perf-test.sh]: Yeah, that's the thing that was especially weird for me, the fact that it was very different to every other tool we had (as you say, because it doesn't use `jopt`). Is it worth filing an issue to use `jopt` so that it works similarly to other command-line tools?

Comment (junrao) [inline bin/kafka-producer-perf-test.sh]: Yes, we should fix jopt before exposing the new ProducerPerformance in bin/. We can either fix the jopt here or in a separate jira.

Comment (ijuma) [inline bin/kafka-producer-perf-test.sh]: Sounds good, @omkreddy what do you prefer?

Comment (omkreddy) [inline bin/kafka-producer-perf-test.sh]: I will fix this here itself. thanks for the reviews.

Comment (omkreddy) [inline bin/kafka-producer-perf-test.sh]: Currently core project uses jopt-simple jar and tools project uses argparse4j jar.  argparse4j is used in VerifiableProducer.  I think we should remove one of the dependency. Can I remove argparse4 dependency and use jopt-simple in tools project?  cc @@junrao  @granders

Comment (granders) [inline bin/kafka-producer-perf-test.sh]: @omkreddy 
I agree the lack of consistency is a bit suspect, but I ran this change by a few committers, and no one seemed to mind. I also didn't see any obvious drawbacks given that it's in its own jar.

As for why I used argparse4j, I think it was roughly:
- partly ease of use and cleaner api, particularly when using in java
- partly curiosity
- partly when writing the java tool, I was unable to get joptsimple to parse -1 properly as an integer as opposed to a flag. This was actually pretty surprising, so maybe I missed something obvious

Comment (omkreddy) [inline bin/kafka-producer-perf-test.sh]: This patch brings kafka-tools.jar to Kafka/libs directory. that means we also need to copy all the dependent jars (argparse4j , jackson-databind).

cc @junrao @ijuma Any suggestions to proceed?

Comment (junrao) [inline bin/kafka-producer-perf-test.sh]: @omkreddy : Discussed this with @granders offline. It seems there is no strong benefit of argparse4 over jopt-simple. For consistency, we can change VerifiableProducer to use jopt-simple for now. We just want to make sure that existing system tests don't break after the change.

Comment (ijuma) [inline bin/kafka-producer-perf-test.sh]: @omkreddy It's a bit unfortunate that we currently run the tools from within the core project. As @junrao said, we can remove the dependency on `argparse4j`, but that still leaves the `jackson-databind` dependency right? Unless we ban any additional dependencies in the `tools` project, it seems to me like there is no choice, but to copy all dependent jars as well.

**Comment by junrao** [comment]
@omkreddy Are you still working on this? If so, perhaps you can include the fix for KAFKA-2516 here too.

**Comment by omkreddy** [comment]
@junrao  Yes, I will do the required changes. Will submit patch shortly.

**Comment by gwenshap** [comment]
@omkreddy any updates? this also needs rebasing

File: bin/kafka-producer-perf-test.sh (hunk 2 above)
Comment (omkreddy) [inline bin/kafka-producer-perf-test.sh]: @junrao @ijuma  After doing some implementation, I felt argparse4j is more flexible than jopt-simple. I have used argparse4j  for parsing command line options. In future, we can migrate existing jopt usages to argparse4j.

**Comment by omkreddy** [comment]
@gwenshap  I just uploaded rebased patch.

File: tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java [hunk 4]
```diff
@@ -3,67 +3,133 @@
...
-        for (int i = 0; i < numRecords; i++) {
-            long sendStartMs = System.currentTimeMillis();
-            Callback cb = stats.nextCompletion(sendStartMs, payload.length, stats);
-            producer.send(record, cb);
-
-            if (throttler.shouldThrottle(i, sendStartMs)) {
-                throttler.throttle();
+        try {
+            Namespace res;
+            res = parser.parseArgs(args);
```
Comment (ijuma) [inline tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java]: The two lines above can be written as a single line.

Comment (omkreddy) [inline tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java]: ok..done the changes

**Comment by gwenshap** [comment]
LGTM. 

I'm merging this and we can resolve the argparse4j vs jopt later. This patch does not add new inconsistencies.

**Comment by gwenshap** [comment]
Sorry, we can't merge. There is a conflict on producer_performance.py. 

@omkreddy, can you rebase?

**Comment by ijuma** [comment]
Quick question: are the option names consistent with the other tools?

**Comment by omkreddy** [comment]
@gwenshap updated PR

@ijuma  replaced "_" character with "-" to make the names consistent with other tools.  There is some inconsistency in option names. For example VerifiableProducer uses the names like "--max--messages" and  ProducerPerformance uses the names like "--num-records",  "--record-size".etc.. I have not changed these names..  If required, I will change.

File: tests/kafkatest/services/performance/producer_performance.py [hunk 5]
```diff
@@ -46,6 +46,7 @@ def __init__(self, context, num_nodes, kafka, security_protocol, topic, num_reco
 
     def _worker(self, idx, node):
         args = self.args.copy()
+<<<<<<< HEAD
```
Comment (gwenshap) [inline tests/kafkatest/services/performance/producer_performance.py]: accidental leftover?

File: tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java [hunk 6]
```diff
@@ -3,67 +3,135 @@
...
-            producer.send(record, cb);
-
-            if (throttler.shouldThrottle(i, sendStartMs)) {
-                throttler.throttle();
+        try {
+            Namespace res = parser.parseArgs(args);
+
+            /* parse args */
+            String topicName = res.getString("topic");
+            long numRecords = res.getLong("numRecords");
```
Comment (gwenshap) [inline tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java]: shouldn't this be the argument name? num-records and record-size?

**Comment by gwenshap** [comment]
Note that Jenkins test is still failing due to merge errors.

I also think there's a leftover "<<< HEAD" in there, may be related.

File: tests/kafkatest/services/performance/producer_performance.py (hunk 5 above)
Comment (omkreddy) [inline tests/kafkatest/services/performance/producer_performance.py]: oops..sorry for this clear error. I have tested the ProducerPerformance code. As i am not ready with my setup, I have not tested system test.

**Comment by omkreddy** [comment]
@gwenshap yes, the arguments are of the form "num-records" ,"record-size".  "-" has special meaning in  argparse4j. So while defining arguments, we need to give unique name  and we can use this unique name to read the arguments .

**Comment by gwenshap** [comment]
LGTM

**Comment by b-goyal** [comment]
Updated  kafka-producer-perf-test.sh to use org.apache.kafka.clients.tools.ProducerPerformance. 
Updated build.gradle to add kafka-tools-0.9.0.0-SNAPSHOT.jar to kafka/libs  folder.
